### PR TITLE
vSphere custom vm params

### PR DESCRIPTION
**Pull request by chrislovecnm**
This provides support for https://github.com/hashicorp/terraform/issues/3834

Adding new map parameter that contains key value pairs that are stored to ExtraConfig.  This provides the capability to store metadata within a virtual machine.

Acceptance test have been provided.

### CODE REVIEW

**Comment by chrislovecnm** [comment]
@phinze might you take a look?  Would appreciate getting this into master and figuring out when it will be released.  Like to coordinate release schedules.

File: CHANGELOG.md [hunk 1]
```diff
@@ -51,6 +51,7 @@ IMPROVEMENTS:
   * provider/digitalocean: Make user_data force a new droplet [GH-3740]
   * provider/vsphere: Do not add network interfaces by default [GH-3652]
   * provider/openstack: Configure Fixed IPs through ports [GH-3772]
+  * provider/openstack: Specify a port ID on a Router Interface [GH-3903]
```
Comment (jen20) [inline CHANGELOG.md]: I don't spot any changes to the openstack provider in this pull request - is this a remnant of another branch? In general we'd prefer it if pull requests didn't update the CHANGELOG, as it tends to lead to frequent merge conflicts. Could you remove this from the PR?

File: builtin/providers/vsphere/resource_vsphere_virtual_machine.go [hunk 2]
```diff
@@ -418,6 +436,19 @@ func resourceVSphereVirtualMachineRead(d *schema.ResourceData, meta interface{})
...
+		//TODO: can only set specific custom value, not everything
+		//Would need the config here
+		//custom_configs := make(map[string]types.AnyType)
+		for _, v := range mvm.Config.ExtraConfig {
+			value := v.GetOptionValue()
+			//custom_configs[value.Key] = value.Value
+			log.Printf("[DEBUG] custom configs %s,%s",value.Key, value.Value)
+		}
+		//d.Set("custom_configuration_parameters", custom_configs)
+	}
```
Comment (jen20) [inline builtin/providers/vsphere/resource_vsphere_virtual_machine.go]: Does this loop do anything beyond printing to the log? It's unclear whether `v.GetOptionValue()` may have any side effects.

**Comment by jen20** [comment]
Hi @chrislovecnm! I've looked over the code and left a few comments inline. I have no vSphere environment here, however, so I'm relying on the acceptance tests being correct.

File: CHANGELOG.md (hunk 1 above)
Comment (chrislovecnm) [inline CHANGELOG.md]: Can do ... it is from updating from the master

File: builtin/providers/vsphere/resource_vsphere_virtual_machine.go (hunk 2 above)
Comment (chrislovecnm) [inline builtin/providers/vsphere/resource_vsphere_virtual_machine.go]: we can simply remove this section.  I was attempting to pull back the custom config values that are being set, but the list is a lot longer that provided.

**Comment by chrislovecnm** [comment]
@jen20 will work on the two comments.  Lots of testing done, and found a couple more bugs when you hit an ipv6 address.

**Comment by chrislovecnm** [comment]
@jen20  so now the build is failing on files that I have not changed ... Joy :smile: 

I addressed your comments.  Lots of testing done

**Comment by chrislovecnm** [comment]
@jen20 merged in upstream master, and now build is happy.  We good with a merge now?

File: builtin/providers/vsphere/resource_vsphere_virtual_machine_test.go [hunk 3]
```diff
@@ -127,6 +127,67 @@ func TestAccVSphereVirtualMachine_dhcp(t *testing.T) {
...
+					resource.TestCheckResourceAttr(
+						"vsphere_virtual_machine.car", "disk.0.template", template),
+					resource.TestCheckResourceAttr(
+						"vsphere_virtual_machine.car", "network_interface.#", "1"),
+					resource.TestCheckResourceAttr(
+						"vsphere_virtual_machine.car", "custom_configuration_parameters.foo", "bar"),
+					resource.TestCheckResourceAttr(
+						"vsphere_virtual_machine.car", "custom_configuration_parameters.car", "ferrai"),
+					resource.TestCheckResourceAttr(
+						"vsphere_virtual_machine.car", "custom_configuration_parameters.num", "42"),
```
Comment (phinze) [inline builtin/providers/vsphere/resource_vsphere_virtual_machine_test.go]: These helpers are only checking the state file, so right now no verification is being done on the custom parameters.

In other words, if you remove all your code but the addition of the field to the schema, this test will still pass!

Something along these lines:

```go
func testAccCheckVSphereVirtualMachineCustomConfigs(vm *virtualMachine) resource.TestCheckFunc {
  return func(s *terraform.State) error {
    // use the vmware client to ensure that the passed in vm has the expected custom config
    // return an error if not
  }
}
```

I'd recommend writing a `CheckFunc` that accepts `&vm` and inspects it to prove the custom attributes get se the

**Comment by phinze** [comment]
Hi @chrislovecnm - looks like the acceptance test could use a bit more work here. Details inline.

Also could you rebase and squash this branch against the latest master? If not we can take care of that fore you at merge time. :+1:

File: builtin/providers/vsphere/resource_vsphere_virtual_machine_test.go (hunk 3 above)
Comment (chrislovecnm) [inline builtin/providers/vsphere/resource_vsphere_virtual_machine_test.go]: This is exactly what I was asking about in another ticket :smile:

**Comment by chrislovecnm** [comment]
@phinze I will try a rebase again.... It turned into one of those rebases :laughing:

**Comment by chrislovecnm** [comment]
@phinze integration test done.  All acceptance tests are passing.  Comments?

**Comment by chrislovecnm** [comment]
@jen20 @phinze it would be good to get this rolled in, my work on my project is getting pushed because of this ... pretty please :smile:

**Comment by phinze** [comment]
Hey @chrislovecnm - sorry for the delay! Taking a look now

**Comment by phinze** [comment]
LGTM!

**Pull request merged**